src/Sylius/Component/Core/Promotion/Action/DiscountAction.php

```diff
@@ -51,34 +44,46 @@ public function __construct(FactoryInterface $adjustmentFactory, OriginatorInter
      */
     public function revert(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion)
     {
-        if (!$subject instanceof OrderInterface && !$subject instanceof OrderItemInterface) {
-            throw new UnexpectedTypeException(
-                $subject,
-                'Sylius\Component\Core\Model\OrderInterface or Sylius\Component\Core\Model\OrderItemInterface'
-            );
+        if (!$this->isSubjectValid($subject)) {
+            return;
         }
 
-        foreach ($subject->getAdjustments(AdjustmentInterface::ORDER_PROMOTION_ADJUSTMENT) as $adjustment) {
-            if ($promotion === $this->originator->getOrigin($adjustment)) {
-                $subject->removeAdjustment($adjustment);
+        foreach ($subject->getItems() as $item) {
+            foreach ($item->getUnits() as $unit) {
+                $this->removeUnitOrderPromotionAdjustmentsByOrigin($unit, $promotion);
             }
         }
     }
 
     /**
-     * @param PromotionInterface $promotion
-     * @param string $type
+     * @param PromotionSubjectInterface $subject
      *
-     * @return AdjustmentInterface
+     * @return bool
      */
-    protected function createAdjustment(PromotionInterface $promotion, $type = AdjustmentInterface::ORDER_PROMOTION_ADJUSTMENT)
+    protected function isSubjectValid(PromotionSubjectInterface $subject)
     {
-        $adjustment = $this->adjustmentFactory->createNew();
-        $adjustment->setType($type);
-        $adjustment->setLabel($promotion->getName());
+        if (!$subject instanceof OrderInterface) {
+            throw new UnexpectedTypeException($subject, OrderInterface::class);
+        }
 
-        $this->originator->setOrigin($adjustment, $promotion);
+        return 0 !== $subject->countItems();
+    }
 
-        return $adjustment;
+    /**
+     * @param array $configuration
+     */
+    protected abstract function isConfigurationValid(array $configuration);
```

Comment:
I would suggest tot put this method at the top of class. Even if its protected, is it pretty important... There is even some [example](http://www.php-fig.org/psr/psr-2/#4-5-abstract-final-and-static) of fig